// @gathertown/gather-game-client/src/TransactionManager.ts
const TXN_TIMEOUT_LENGTH_MS = 30000;

import { TransactionStatus } from "@gathertown/gather-game-common/dist/src/public/events";
import { logger } from "./Logger";

export class TransactionManager {
  private pendingTxns: {
    [txnId: number]: { res: (succeeded: boolean) => void; rej: (reason: string) => void };
  } = {};
  private txnTimeouts: { [txnId: number]: ReturnType<typeof setTimeout> } = {};

  addTransaction(useLongerTimeout?: boolean) {
    const timeoutLength = useLongerTimeout ? TXN_TIMEOUT_LENGTH_MS * 2 : TXN_TIMEOUT_LENGTH_MS;
    const txnId = this.getRandomTxnId();
    const txnPromise = new Promise((res, rej) => {
      if (txnId === undefined) {
        return;
      }
      this.pendingTxns[txnId] = { res, rej };
      this.txnTimeouts[txnId] = setTimeout(() => {
        // @ts-expect-error Error auto-ignored when enabling TS noUncheckedIndexedAccess. If you're already touching this code, please clean this up while you're at it.
        // TODO: @ENG-4257 Clean these up! See the linear task for more context and advice for cleaning up.
        this.pendingTxns[txnId].rej("Transaction timed out");
      }, timeoutLength);
    });
    return { txnId, txnPromise };
  }

  rejectTransaction(txnId: number, message?: string) {
    this.pendingTxns[txnId]?.rej(message ?? "Transaction was rejected");
    // @ts-expect-error Error auto-ignored when enabling TS noUncheckedIndexedAccess. If you're already touching this code, please clean this up while you're at it.
    // TODO: @ENG-4257 Clean these up! See the linear task for more context and advice for cleaning up.
    clearTimeout(this.txnTimeouts[txnId]);
    delete this.pendingTxns[txnId];
    delete this.txnTimeouts[txnId];
  }

  handleTransactionStatusEvent(transactionStatus: TransactionStatus) {
    const { txnId, succeeded, reason } = transactionStatus;
    if (this.pendingTxns[txnId] === undefined) {
      logger.error("Received a txnId for a non-pending transaction.");
      return;
    }
    if (succeeded) {
      // @ts-expect-error Error auto-ignored when enabling TS noUncheckedIndexedAccess. If you're already touching this code, please clean this up while you're at it.
      // TODO: @ENG-4257 Clean these up! See the linear task for more context and advice for cleaning up.
      this.pendingTxns[txnId].res(true);
    } else {
      // @ts-expect-error Error auto-ignored when enabling TS noUncheckedIndexedAccess. If you're already touching this code, please clean this up while you're at it.
      // TODO: @ENG-4257 Clean these up! See the linear task for more context and advice for cleaning up.
      this.pendingTxns[txnId].rej(`Transaction failed due to error. ${reason ?? ""}`);
    }
    // @ts-expect-error Error auto-ignored when enabling TS noUncheckedIndexedAccess. If you're already touching this code, please clean this up while you're at it.
    // TODO: @ENG-4257 Clean these up! See the linear task for more context and advice for cleaning up.
    clearTimeout(this.txnTimeouts[txnId]);
    delete this.pendingTxns[txnId];
    delete this.txnTimeouts[txnId];
  }

  reset() {
    Object.values(this.txnTimeouts).forEach((timeout) => clearTimeout(timeout));
    Object.values(this.pendingTxns).forEach(({ rej }) =>
      rej("Transaction failed due to forced reset, likely because the connection failed"),
    );
    this.pendingTxns = {};
    this.txnTimeouts = {};
  }

  // txnId is a uint32 so max possible is 2^32 - 1 = 4294967295
  private getRandomTxnId() {
    return Math.floor(Math.random() * 4294967295);
  }
}
